lib/opte/src/engine/layer.rs

```diff
@@ -1402,12 +1503,33 @@ impl Layer {
     ///
     /// Updating the ruleset immediately invalidates all flows
     /// established in the Flow Table.
-    pub(crate) fn set_rules(
+    pub fn set_rules(
         &mut self,
         in_rules: Vec<Rule<Finalized>>,
         out_rules: Vec<Rule<Finalized>>,
     ) {
         self.ft.clear();
+        self.set_rules_core(in_rules, out_rules);
+    }
+
+    /// Set all rules at once without clearing the flow table.
+    ///
+    /// See [FlowTable::mark_dirty] for the performance and correctness
+    /// implications.
+    pub fn set_rules_soft(
```

Comment:
We could, with the current design, just revalidate all actions at the time `set_rules` is called. The lazy approach was partly designed around possibly needing access to packet metadata and body for action validation down the line.